Lib/asyncio/controller.py

```diff
@@ -0,0 +1,78 @@
+"""Protocol server controller."""
+
+__all__ = ['Controller']
+
+
+import asyncio
+import os
+import socket
+import threading
+
+try:
+    from socket import socketpair
+except ImportError:
+    from asyncio.windows_utils import socketpair
+
+
+class Controller:
+    def __init__(self, hostname, port, ready_timeout=1.0, loop=None):
+        self.hostname = hostname
+        self.port = port
+        self.loop = asyncio.new_event_loop() if loop is None else loop
+        self.server = None
+        self._thread = None
+        self._thread_exception = None
+        envar = os.getenv('PYTHONASYNCIOCONTROLLERTIMEOUT')
+        self.ready_timeout = ready_timeout if envar is None else float(envar)
+        # For exiting the loop.
+        self._rsock, self._wsock = socketpair()
+        self.loop.add_reader(self._rsock, self._reader)
```

Comment:
I'm not sure that a socketpair is really needed here. Why not simply using call_soon_threadsafe() from stop?